File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

class Empty { }                              // Noncompliant {{Remove this empty class, write its code or make it an "interface".}}
//    ^^^^^

public class PublicEmpty { }                 // Noncompliant
internal class InternalEmpty { }             // Noncompliant

class EmptyWithComments                      // Noncompliant
{
    // Some comment
}

class ClassWithProperty
{
    int SomeProperty => 42;
}
class ClassWithField
{
    int SomeField = 42;
}
class ClassWithMethod
{
    void Method() { }
}
class ClassWithIndexer
{
    int this[int index] => 42;
}
class ClassWithDelegate
{
    delegate void MethodDelegate();
}
class ClassWithEvent
{
    event EventHandler CustomEvent;
}

class OuterClass
{
    class InnerEmpty1 { }                    // Noncompliant
    private class InnerEmpty2 { }            // Noncompliant
    protected class InnerEmpty3 { }          // Noncompliant
    internal class InnerEmpty4 { }           // Noncompliant
    protected internal class InnerEmpty5 { } // Noncompliant
    public class InnerEmpty6 { }             // Noncompliant

    public class InnerEmptyWithComments      // Noncompliant
    {
        // Some comment
    }

    class InnerNonEmpty
    {
        public int SomeProperty => 42;
    }
}

class GenericEmpty<T> { }                    // Noncompliant
//    ^^^^^^^^^^^^
class GenericEmptyWithConstraints<T>         // Noncompliant
    where T : class
{
}

class GenericNotEmpty<T>
{
    void Method(T arg) { }
}
class GenericNotEmptyWithConstraints<T>
    where T : class
{
    void Method(T arg) { }
}

class IntegerList: List<int> { }             // Compliant - creates a more specific type without adding new members to it (similar to using the typedef keyword in C/C++)
class StringLookup<T>: Dictionary<string, T> { }
interface IIntegerSet<T>: ISet<int> { }

[ComVisible(true)]
class ClassWithAttribute { }                 // Compliant - types with attributes are ignored

static class StaticEmpty { }                 // Noncompliant

abstract class AbstractEmpty { }             // Noncompliant

partial class PartialEmpty { }               // Compliant - Source Generators and some frameworks use empty partial classes as placeholders

partial class PartialNotEmpty
{
    int Prop => 42;
}

interface IMarker { }                        // Compliant - this rule only deals with classes

struct EmptyStruct { }                       // Compliant - this rule only deals with classes

enum EmptyEnum { }                           // Compliant - this rule only deals with classes

class Conditional                            // Compliant - it's not empty when the given symbol is defined
{
#if NOTDEFINED
    public override string ToString()
    {
        return "Debug Text";
    }
#endif
}

class { }                                    // Error

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ClassShouldNotBeEmpty.vb
Imports System

Class Empty                                         ' Noncompliant {{Remove this empty class, or add members to it.}}
    ' ^^^^^
End Class


Public Class PublicEmpty                            ' Noncompliant
End Class

Class InternalEmpty                                 ' Noncompliant
End Class

Class EmptyWithComments                             ' Noncompliant
    ' Some comment
End Class

Class ClassWithProperty
    Private ReadOnly Property SomeProperty As Integer
        Get
            Return 42
        End Get
    End Property
End Class

Class ClassWithField
    Private SomeField As Integer = 42
End Class

Class ClassWithMethod
    Private Sub Method()
    End Sub
End Class

Class ClassWithIndexer
    Private ReadOnly Property Item(index As Integer) As Integer
        Get
            Return 42
        End Get
    End Property
End Class

Class ClassWithDelegate
    Delegate Sub MethodDelegate()
End Class

Class ClassWithEvent
    Private Event CustomEvent As EventHandler
End Class

Class OuterClass
    Class InnerEmpty1                               ' Noncompliant
    End Class

    Private Class InnerEmpty2                       ' Noncompliant
    End Class

    Protected Class InnerEmpty3                     ' Noncompliant
    End Class

    Class InnerEmpty4                               ' Noncompliant
    End Class

    Protected Class InnerEmpty5                     ' Noncompliant
    End Class

    Public Class InnerEmpty6                        ' Noncompliant
    End Class

    Public Class InnerEmptyWithComments             ' Noncompliant
        ' Some comment
    End Class

    Class InnerNonEmpty
        Public ReadOnly Property SomeProperty As Integer
            Get
                Return 42
            End Get
        End Property
    End Class
End Class

Class GenericEmpty(Of T)
    ' ^^^^^^^^^^^^
End Class

Class GenericEmptyWithConstraints(Of T As Class)    ' Noncompliant
End Class

Class GenericNotEmpty(Of T)
    Private Sub Method(arg As T)
    End Sub
End Class

Class GenericNotEmptyWithConstraints(Of T As Class)
    Private Sub Method(arg As T)
    End Sub
End Class

MustInherit Class AbstractEmpty                     ' Noncompliant
End Class

Partial Class PartialEmpty                          ' Compliant - Source Generators and some frameworks use empty partial classes as placeholders
End Class

Partial Class PartialNotEmpty
    Public ReadOnly Property Prop As Integer
        Get
            Return 42
        End Get
    End Property
End Class

Interface IMarker                                   ' Compliant - this rule only deals with classes
End Interface

Structure EmptyStruct                               ' Compliant - this rule only deals with classes
End Structure

Class                                               ' Error
End Class                                    
